### DISCUSSION TITLE
How to build entirely in docker?

### DISCUSSION THREAD

**Discussion by xrd** [pocketbase/pocketbase#2621] [Q&A] [answered]
I build a framework app a long time ago (version 0.7.4) and got it all working and building inside docker. At that point, I believe I generated a go.mod and go.sum file using my local go toolchain. However, I've forgotten how to regenerate those files and now want to update to the newest version of pocketbase.

My preference would be that all that happens inside the dockerfile, rather than relying on my failing memory to remember what tools I installed, what paths I was in (I recall go cares a lot about the file system structure), etc. Dockerfiles are so great for capturing build steps in a declarative way, and I would happily sacrifice some degree of uncertainty about the exact versions (perhaps with an escape hatch where I can dump the go.sum file (?) once the steps are down and version control that file). 

I realize this is more a golang question that specifically about pocketbase, but not having to manage the golang toolchain and understand all the commands would, IMHO, make this more accessible to people. Months ago I was able to build a framework app without knowing go at all, and that is so amazing.

For example, this link has the main.go, but not the rest of the commands needed to complete the build.

https://pocketbase.io/docs/use-as-framework/

**Comment by ganigeorgiev** [comment, marked as answer]
If you haven't already, you can check the official Docker Go guide - https://docs.docker.com/language/golang/build-images/.

In general, to update all your Go modules you can run:

```sh
$ go get -u   # update all modules listed in go.mod
$ go mod tidy # adds any missing imported modules and remove unnecessary ones
```

The files `go.mod` and `go.sum` usually should be part of version control, aka. they need to be `COPY`-ied in the image similar to your source Go files. You have to threat them the same way as `package.json`, `pubspec.yaml`. `composer.json`, etc. similar deps management files.

---

> My preference would be that all that happens inside the dockerfile, rather than relying on my failing memory to remember what tools I installed, what paths I was in (I recall go cares a lot about the file system structure), etc.

If you don't care about the exact used versions and want to generate the deps files when building the image, your Dockerfile could look something like:

```dockerfile
FROM golang:latest

WORKDIR /src

# copy all your .go source files
# (or use a .dockerignore and COPY . .)
COPY *.go .

# remove any previously initialized go.mod and go.sum files
# (this is in case the container data wasn't destroyed)
RUN rm -f go.mod rm go.sum

# initialize Go modules
RUN go mod init app

# fetch dependencies
RUN go mod tidy

# build (switch to 1 to use the CGO SQLite)
RUN CGO_ENABLED=0 go build -o /pocketbase

# export listener port
EXPOSE 8090

# run
# (ps. don't forget to mount the pb_data as volume in /pb_data)
CMD ["/pocketbase", "serve", "--http=0.0.0.0:8090"]
```
_(optionally you can create a multi-stage image to separate the build process)_

But I wouldn't recommend the above as it will be very slow (every time when you add a new dependency you'll have to recreate the image).

If I had to use Docker, I'd create the go.mod file manually and `COPY` it similarly to the guide from the Docker site.

**Reply by xrd** [reply]
This looks great and works for me.

I might try to create a dockerfile which runs the ` go mod init app` and `go mod tidy` steps only if the go.sum and go.mod files are missing. That way I can generate them once the first time I build (without installing golang on my local machine) and then version control them and update only when necessary. I think this will be a good middle ground for a golang newbee like me. 

Thanks so much!

**Comment by pnmcosta** [comment]
Here's an example of a multi-stage image build targeting `alpine` with CGO_ENABLED=1 that I use:

```
# Cache gogcc alpine
FROM golang:1.20-alpine as gogcc
ENV GOOS=linux 
ENV GOARCH=amd64 
ENV CGO_ENABLED=1

RUN apk update && apk add --no-cache \
        gcc \
        musl-dev

# Build the binary
FROM gogcc as builder

WORKDIR /app

# Download dependencies
COPY go.mod .
COPY go.sum .
RUN go mod download && go mod verify

# Build /app/bin 
COPY internal ./internal
COPY migrations ./migrations
COPY main.go .

RUN go build -ldflags="-s -w" -o bin -v ./main.go

# Serve the binary with pb_public
FROM alpine:latest as bin

RUN apk update && apk add --no-cache \
        gcc \
        musl-dev

WORKDIR /app/
COPY pb_public ./pb_public
COPY --from=builder /app/bin .

EXPOSE 8080

CMD ["/app/bin", "serve", "--http=0.0.0.0:8080"]
```

**Reply by xrd** [reply]
Hi @pnmcosta thanks for sharing this! Do you generate the go.mod and go.sum file before you build inside docker? I'm not very familiar with golang build steps and get confused with what starting point I need. Yesterday I tried to run `go mod download` with only a go.mod file that contained `module pocketbase` and the golang build quickly reminded me that `github.com/pocketbase/dbx` was missing, so I needed to get that (marked in that file as indirect). This is where I get stuck, I'm unclear what dependencies I need and which ones to choose. I appreciate you adding to the conversation!

**Reply by ganigeorgiev** [reply]
> I'm unclear what dependencies I need and which ones to choose

You can run `go mod tidy`. It will update the go.mod with any imported dependencies that are missing and remove the unnecessary ones.

**Reply by pnmcosta** [reply]
Sort of, `go.mod` and `go.sum` are like `package.json` and `package-lock.json` from nodejs stack. I suggest reading https://go.dev/blog/using-go-modules, same workflow should apply to PB as framework.
